firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/core/CrashlyticsController.java

```diff
@@ -1805,26 +1809,84 @@ private void sendSessionReports(AppSettingsData appSettings, boolean dataCollect
     }
   }
 
-  private void recordFatalFirebaseEvent(long timestamp) {
-    if (firebaseCrashExists()) {
+  /**
+   * Helper class that listens for Crashlytics origin events from FA and provides a method to block
+   * while waiting for a Crash event to occur. The implementation assumes there can be at most ONE
+   * app exception event being processed and waited upon at a time.
+   */
+  private static class BlockingCrashEventListener
+      implements AnalyticsReceiver.CrashlyticsOriginEventListener {
+
+    private final CountDownLatch eventLatch = new CountDownLatch(1);
+
+    public void awaitEvent() throws InterruptedException {
```

Comment:
As mentioned in chat, if you used a `TaskCompletionSource` instead of a countdown latch above, returned its `Task` from this method, and just resolved the task (via the completion source) in `onCrashOriginEvent`, you could simplify a lot of what's downstream.

1) You wouldn't need the extra `Tasks.call` with its own executor down below because this would just become part of the task chain already in the exception handler.
2) You wouldn't need the extra 2-second timeout here because the timeout would be handled by the exception handler task chain's timeout.

Reply:
(Going to do this on a future PR. See previous comment.)